neps/nep-0452.md

```diff
@@ -0,0 +1,221 @@
+/// choose what should happen. For example, the contract can choose to keep the assets 
+/// or send them back to the original linkdrop creator.
+/// 
+/// Requirements:
+/// * The predecessor account *MUST* be the current contract ID.
+/// * The assets being sent *MUST* be associated with the signer’s public key.
+/// * The assets *MUST* be sent to the `account_id` passed in.
+///
+/// Arguments:
+/// * `account_id` the account that should receive the linkdrop assets.
+///
+/// Returns `true` if the claim was successful meaing all assets were sent to the `account_id`.
+function claim(account_id: string) -> Promise<boolean>
+
+
+
+/// Creates a new NEAR account and transfers all assets linked to the signer’s public key to 
+/// the *newly created account*. If the transfer fails for whatever reason, it is up to the 
+/// smart contract developer to choose what should happen. For example, the contract can    
+/// choose to keep the assets or return them to the original linkdrop creator. 
+///
+/// Requirements 
+/// * The predecessor account *MUST* be the current contract ID.
+/// * The assets being sent *MUST* be associated with the signer’s public key.
+/// * The assets *MUST* be sent to the `new_account_id` passed in.
+/// * The newly created account *MUST* have a new access key added to its account (either       
+///   full or limited access) in the same receipt that the account was created in. 
+/// 
+/// Arguments
+/// * `new_account_id`: the valid NEAR account which is being created and should 
+///   receive the linkdrop assets
+/// * `new_public_key`: the valid public key that should be used for the access key added to         
+/// The Public key must be in a binary format with base58 string serialization with human-readable curve. The key types currently supported are secp256k1 and ed25519. Ed25519 public keys accepted are 32 bytes and secp256k1 keys are the uncompressed 64 format.
+///   the newly created account.
+///
+/// Returns `true` if the claim was successful meaning the `new_account_id` was created and all assets were sent to it.
+function create_account_and_claim(new_account_id: string, new_public_key: string) -> Promise<boolean>
+```
+
+## Example Scenarios
+
+*Pre-requisite Steps*: Linkdrop creation:
+The linkdrop creator that has an account and some $NEAR:
+
+- creates a keypair locally (`pubKey1`, `privKey1`). (The keypair is not written to chain at this time)
+- calls a method, passing the `pubKey1` and desired $NEAR amount as arguments, to a contract that implements the linkdrop standard.
+- The contract will map the `pubKey1` to the desired balance for the linkdrop.
+- The contract will then add the `pubKey1` as a function call access key with the ability to call `claim` and `create_account_and_claim`. This means that anyone with the `privKey1` (see above), can sign a transaction on behalf of this contract (signer id set to contract id) with a function call to call one of the mentioned functions to claim the assets.
```

Comment:
I wonder is this restriction really needed? I think you can check the public key from inside the contract using the function `signer_account_pk` (I haven't tried it though to be honest).

Reply:
@fadeevab Could you suggest how to phrase it less restrictive? Given it is an example, I don't feel it should be taken as a requirement, but it is definitely the way it is currently implemented on Wallets that create `*.near` accounts via linkdrops.

P.S. `signer_account_pk` is used in the implementation: https://github.com/near/near-linkdrop/blob/49279e529c254fa7736465b4a39d05cb8f1e5443/src/lib.rs#L92

Reply:
@frol From my understanding, it relates to another line "A user ..... signs a transaction on behalf of the linkdrop contract (contract is set to signer_id) using privKey1. The method will be create_account_and_claim." There is no explanation why, but I think it's like approval of a linkdrop by the contract creator: the contract owner must add every pubKey1 to the contract account. I believe it's not necessary, also, it will force linkdrop creators to do extra steps of adding every pubKey1 to the account keeping their keys in warm wallets.

Until there is a clear explanation why it's needed, I would remove "on behalf of the linkdrop contract (contract is set to signer_id)", also I would remove the whole paragragh "The contract will then add the `pubKey1` as a function call access key ........ ". It would essentially remove the following assert https://github.com/near/near-linkdrop/blob/49279e529c254fa7736465b4a39d05cb8f1e5443/src/lib.rs#L81.

Again, tell me, if there is misunderstanding from my side. Thanks!

Reply:
@BenKurrek I think @fadeevab missed the following fact (which indicates that the current NEP is not explicit enough):

The original use-case for the linkdrops is to on-board users who DO NOT HAVE an account on the blockchain. Thus, they cannot sign a transaction on their own behalf. Linkdrop link provides user with an access key that belongs to some existing account (linkdrop contract, e.g. `near` account on mainnet), and so they can construct a transaction on behalf of the linkdrop account and sign it with the provided limited access key.

@fadeevab Does it clarify the situation for you?

Reply:
@frol Yes, thanks 👍 I believe this point is crusial and must be indicated explicitly.

Reply:
I've added some background to the top of the NEP